File: rules/benchmarks/spikes/mmap/process-accounting/exploratory/launch.c
// Copyright 2026 Google LLC
// SPDX-License-Identifier: Apache-2.0
#include <errno.h>
#include <fcntl.h>
#include <spawn.h>
#include <stdint.h>
#include <stdio.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>
extern char **environ;
static uint64_t now(clockid_t clock) {
    struct timespec t;
    if (clock_gettime(clock, &t)) _exit(125);
    return (uint64_t)t.tv_sec * 1000000000 + t.tv_nsec;
}
int main(int argc, char **argv) {
    if (argc < 4) return 125;
    posix_spawn_file_actions_t actions;
    if (posix_spawn_file_actions_init(&actions)) return 125;
    int out = open(argv[1], O_WRONLY|O_CREAT|O_TRUNC, 0600);
    int err = open(argv[2], O_WRONLY|O_CREAT|O_TRUNC, 0600);
    if (out < 0 || err < 0) return 125;
    if (posix_spawn_file_actions_adddup2(&actions, out, STDOUT_FILENO)) return 125;
    if (posix_spawn_file_actions_adddup2(&actions, err, STDERR_FILENO)) return 125;
    if (posix_spawn_file_actions_addclose(&actions, out)) return 125;
    if (posix_spawn_file_actions_addclose(&actions, err)) return 125;
    pid_t child;
    uint64_t start_unix = now(CLOCK_REALTIME), start = now(CLOCK_MONOTONIC);
    int error = posix_spawn(&child, argv[3], &actions, NULL, &argv[3], environ);
    uint64_t spawned = now(CLOCK_MONOTONIC);
    if (error) { errno = error; perror("posix_spawn"); return 125; }
    int status;
    while (waitpid(child, &status, 0) < 0) if (errno != EINTR) return 125;
    uint64_t end = now(CLOCK_MONOTONIC), end_unix = now(CLOCK_REALTIME);
    posix_spawn_file_actions_destroy(&actions);
    close(out); close(err);
    printf("{\"start_unix_ns\":%llu,\"end_unix_ns\":%llu,\"elapsed_ns\":%llu,\"spawn_call_ns\":%llu,\"status\":%d}\n",
           (unsigned long long)start_unix, (unsigned long long)end_unix,
           (unsigned long long)(end-start), (unsigned long long)(spawned-start), status);
    return WIFEXITED(status) ? WEXITSTATUS(status) : 125;
}

File: rules/benchmarks/spikes/mmap/process-accounting/exploratory/launch.rs
// Copyright 2026 Google LLC
// SPDX-License-Identifier: Apache-2.0
use std::{fs::File, process::{Command, Stdio}, time::{Instant, SystemTime, UNIX_EPOCH}};
fn unix() -> u128 { SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos() }
fn main() {
    let args: Vec<_> = std::env::args_os().skip(1).collect();
    let runs = std::env::var("MAGIKA_LAUNCH_RUNS").unwrap_or("1".into()).parse::<usize>().unwrap();
    for repetition in 0..runs {
    let mut command = Command::new(&args[2]);
    command.args(&args[3..]).stdin(Stdio::null())
        .stdout(File::create(&args[0]).unwrap()).stderr(File::create(&args[1]).unwrap());
    let start_unix = unix();
    let start = Instant::now();
    let mut child = command.spawn().unwrap();
    let spawn_ns = start.elapsed().as_nanos();
    let status = child.wait().unwrap();
    let elapsed_ns = start.elapsed().as_nanos();
    let end_unix = unix();
    let trace = std::fs::read_to_string(&args[1]).unwrap();
    let trace = if trace.trim().is_empty() { "null" } else { trace.trim() };
    println!("{{\"repetition\":{repetition},\"trace\":{trace},\"start_unix_ns\":{start_unix},\"end_unix_ns\":{end_unix},\"elapsed_ns\":{elapsed_ns},\"spawn_call_ns\":{spawn_ns}}}");
    assert!(status.success());
    }
}
